src/Transitions/Transition.js

```diff
@@ -0,0 +1,173 @@
+    let enter = this.props.in;
+
+    if (needsTransition) {
+      this.needsTransition = false;
+      this[enter ? 'performEnter' : 'performLeave']();
+    }
+  }
+
+  performEnter() {
+    let node = React.findDOMNode(this);
+
+    this.props.onEnter(node);
+
+    this.setState({ in: true, transitioning: true }, ()=> {
+      this.props.onEntering(node);
+
+      ensureTransitionEnd(node, () => {
+        if ( this.state.in ){
+          this.setState({
+            transitioning: false
+          }, () => this.props.onEntered(node));
+        }
+
+      }, this.props.duration);
+    });
+  }
+
+  performLeave() {
+    let node = React.findDOMNode(this);
+
+    node = this.props.transitioningNode(node) || node;
+
+    this.props.onExit(node);
+
+    this.setState({ in: false, transitioning: true }, () => {
+      this.props.onExiting(node);
+
+      ensureTransitionEnd(node, () => {
+        if ( !this.state.in ){
+          this.setState({ transitioning: false }, ()=> this.props.onExited(node));
+        }
+      }, this.props.duration);
+    });
+  }
+
+  render() {
+    let childProps = omit(this.props, Object.keys(Transition.propTypes).concat('children'));
+
+    let child = this.props.children;
+    let out = !this.state.in && !this.state.transitioning;
+
+    if ( !child || (this.props.unmountOnExit && out) ){
+      return null;
+    }
+
+    let classes = '';
+
+    // for whatever reason classnames() doesn't actually work here, maybe because they aren't singgle classes?
+    if (this.state.in && !this.state.transitioning) {
+      classes += this.props.enterClassName;
+    }
+
+    if (this.state.in && this.state.transitioning) {
+      classes += this.props.enteringClassName;
+    }
+
+    if (!this.state.in && !this.state.transitioning) {
+      classes += this.props.exitClassName;
+    }
+
+    if (!this.state.in && this.state.transitioning) {
+      classes += this.props.exitingClassName;
+    }
+
+    return React.cloneElement(child, {
+      ...childProps,
+      className: classnames(
+          child.props.className
+        , this.props.className
+        , classes)
+    });
+  }
+}
+
+Transition.propTypes = {
+  in:                React.PropTypes.bool,
+  unmountOnExit:     React.PropTypes.bool,
+  duration:          React.PropTypes.number,
+
+  exitClassName:     React.PropTypes.string,
+  exitingClassName:  React.PropTypes.string,
+  enterClassName:    React.PropTypes.string,
+  enteringClassName: React.PropTypes.string,
+
+  transitioningNode: React.PropTypes.func,
+  onEnter:     React.PropTypes.func,
+  onEntering: React.PropTypes.func,
+  onEntered:  React.PropTypes.func,
+  onExit:     React.PropTypes.func,
+  onExiting:  React.PropTypes.func,
+  onExited:   React.PropTypes.func
+};
+
+Transition.defaultProps = {
+  in:       false,
+  duration: 300,
+  transitioningNode: ()=>{},
+  onEnter:    ()=>{},
+  onEntering: ()=>{},
+  onEntered:  ()=>{},
+
+  onExit:     ()=>{},
+  onExiting:  ()=>{},
+  onExited:   ()=>{}
```

Comment:
I thought that here would be possible perf concerns because fat arrow binds all scope,
but as it turns out, `babel` is smart enough here. 

It transpiles this into:
`onExited: function onExited() {},`